src/Components/Components/src/EventHandlerInvokerFactory.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    public sealed class EventHandlerInvokerFactory
+    {
+        public Action CreateDelegate(object receiver, Action action)
+        {
+            if (receiver == null)
+            {
+                throw new ArgumentNullException(nameof(receiver));
+            }
+
+            if (action == null)
+            {
+                throw new ArgumentNullException(nameof(action));
+            }
+
+            // Determine whether the action is already a bound-delegate attached to handler.
+            //
+            // This is important for a case like the following:
+            //
+            //  @* in Outer.cshtml *@
+            //  <Inner OnClick="@OnClick" />
+            //  @functions {
+            //      void OnClick() { ... }
+            //  }
+            //
+            //  @* in Inner.cshtml *@
+            //  <button onclick="@OnClick">Click Me!</button>
+            //  @functions {
+            //      [Parameter] Action OnClick { get; set; }
+            //  }
+            //
+            // We want to make sure to call StateHasChanged on the Outer component instead
+            // of just the Inner component. We can't rely on action.Target to always point
+            // to the outer component because in the case of a non-capturing lambda it won't.
+            if (receiver is IHandleStateChange handler && !object.ReferenceEquals(receiver, action.Target))
+            {
+                return (Action)(() =>
+                {
+                    action();
+                    _ = handler.HandleStateChangeAsync(Task.CompletedTask);
```

Comment:
Note also that this is kinda fire-and-forget. I'm not sure it makes much sense for this method to return a task, because in many many cases we won't be able to ever await it.

Reply:
I think all the above methods need to have Func<Task> based overloads so that we can bind async handlers and flow the task all the way up to the renderer decide what to do

Reply:
Yes, that will be done as well.

Reply:
@javiercn - note that in come cases these event handlers are fire-and-forget. Nothing awaits them and nothing can.